---
title: "GEOsuppl analysis of Series GSE112049"
author: "Hyun Yong Jin"
date: "April 26, 2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```



```{r GSE112049}
#Gautam et al., Nature Immunology 2019
#RNAseq data was linked to the paper, GSE112049
#Look up Notch target genes in the dataset

installed.packages("GEOquery")
#install.packages("GEOquery") #Not available for 3.5.1
R.Version() #current version is 3.5.1, 2018-07-02


#To install for R version 3.5
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("GEOquery", version = "3.8")

library(GEOquery)

getGEOSuppFiles('GSE112049', fetch_files = FALSE) #Check to see what's in the supplement files

getGEOSuppFiles('GSE112049')
tab <- read.delim("GSE112049/GSE112049_rnaseq_matrix.txt.gz")


library(tidyverse)
tib <-as.tibble(tab) #Let's convert it to tibble format
tib
summary(tib)

lin <- function(x, na.rm = FALS) (2^x) #Function to linerize each values

tiblin <- tib %>% mutate_at(vars(starts_with("W")), lin) %>% mutate_at(vars(starts_with("K")), lin) #Convert log value to linear when header starts with W (for WT) or K (for KO)
summary(tiblin)


tibWT <- select(tiblin, starts_with("W"))
tibKO <- select(tiblin, starts_with("K"))

tibFC <- tiblin %>% 
    mutate(KO.mean = rowMeans(tibKO)) %>%
    mutate(WT.mean = rowMeans(tibWT)) %>%
    mutate(FC = KO.mean/WT.mean)

library(broom) #For function tidy

#for-loop t-test
testresults <- vector("list", nrow(tiblin)) #To make empty vector

for (j in seq(nrow(tiblin))) {
  testresults[[j]] <- tidy(t.test(as.data.frame(tibKO[j,]), as.data.frame(tibWT[j,])))
}
  
head(testresults)

t_stats = do.call(rbind, testresults) #list to dataframe

big_data <-bind_cols(tibFC, t_stats)
glimpse(big_data)

head(big_data)

write.csv(big_data, file="GSE112049_all_data.csv") #write all data 


#Select genes with high fold changes
signif <- big_data %>%
  select(Gene, KO.mean, WT.mean, FC, p.value) %>%
  filter(p.value<0.01 & (FC>2 | FC<1/2)) %>%
  arrange(desc(FC))

signif

write.csv(signif, file="GSE112049_significant_2fold_only.csv")



#Focus on Notch and other interesting genes, align based on FC
#There should be more elegant ways of doing it.   


selectgenes <- 
  big_data %>%
  select(Gene, KO.mean, WT.mean, FC, p.value) %>%
  filter(Gene == "Notch1"| Gene=="Notch2" | Gene=="Hes1" | Gene=="Hes2" | Gene=="Nrarp"| Gene=="Dtx1"| Gene == "Rbpj" |Gene =="Actb" | Gene=="Id3"| Gene=="Foxo1" | Gene=="Tcf7" | Gene=="Cxcr5" | Gene=="Il7r" | Gene=="Ctla4" | Gene=="pcdc1" | Gene=="Havcr2" | Gene=="Klrg1" | Gene=="Zeb2" | Gene=="Bcl2" | Gene=="Myb") %>%
  arrange(desc(FC))

selectgenes

write.csv(selectgenes, file="GSE112049_manually_selected_genes.csv")

#Bcl2 and Tcf7 downregulation of Bcl2 upregulatoin in KO cells were recapitulated but Myb levels were even higher in KO cells. 
#Myb floxed allele is targeted for exon II only. Likely detection of non-funtional Myb transcript accumulated in KO. Bender et al., NI 2004 


```
