---
title: "TissueEnrichment"
author: "Hyun Yong Jin"
date: "August 27, 2018"
output:
  html_document: default
  pdf_document: default
---



```{r}
#Generate Tissue Specific Distribution of Genes of Interest from Single Cells RNA Seq


#Start from C4-CD8-LAYN's 274 list of genes
setwd("~/") #Somehow you should directly 

#Among 274 genes one gene is not unique. Turns out two NA values (No gene names assigned)

#Testing Dublicated genes
#> anyDuplicated(inputGenes)
#[1] 168



library(TissueEnrich)

#Based on this , remove two genes with no gene names. Makes 272 genes in total

c4 <- read.csv("c4-cd8-layn-genename.csv", header=FALSE) 
inputGenes <- as.vector(c4[,'V1']) #convert table into integer or character


gs <- GeneSet(geneIds=inputGenes, organism="Homo Sapiens", geneIdType=SymbolIdentifier())

output <- teEnrichment(inputGenes = gs) #Experiment is done! #Given a gene list as input, this function calculates the tissue-specific gene enrichment using tissue-specific genes from either human or mouse RNA-Seq datasets.



```




## Generate plots

```{r plot, echo=FALSE}


library(tidyr)


#Overall tissuse distribution of 274 genes

#Y-axis based on stat
seEnrichmentOutput<-output[[1]]
enrichmentOutput<-setNames(data.frame(assay(seEnrichmentOutput),row.names = rowData(seEnrichmentOutput)[,1]), colData(seEnrichmentOutput)[,1])
enrichmentOutput$Tissue<-row.names(enrichmentOutput)
ggplot(enrichmentOutput,aes(x=reorder(Tissue,-Log10PValue),y=Log10PValue,label = Tissue.Specific.Genes,fill = Tissue))+
      geom_bar(stat = 'identity')+
      labs(x='', y = '-LOG10(P-Value)')+
      theme_bw()+
      theme(legend.position="none")+
      theme(plot.title = element_text(hjust = 0.5,size = 20),axis.title = element_text(size=15))+
      theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),panel.grid.major= element_blank(),panel.grid.minor = element_blank())


#Figure Based on detected gene numbers, Header can be found here
assay(seEnrichmentOutput)

#Y-axis as gene numbers
seEnrichmentOutput<-output[[1]]
enrichmentOutput<-setNames(data.frame(assay(seEnrichmentOutput),row.names = rowData(seEnrichmentOutput)[,1]), colData(seEnrichmentOutput)[,1])
enrichmentOutput$Tissue<-row.names(enrichmentOutput)
ggplot(enrichmentOutput,aes(x=reorder(Tissue,-Tissue.Specific.Genes),y=Tissue.Specific.Genes,label = Tissue.Specific.Genes,fill = Tissue))+
      geom_bar(stat = 'identity')+
      labs(x='', y = 'Numbers of Genes')+
      theme_bw()+
      theme(legend.position="none")+
      theme(plot.title = element_text(hjust = 0.5,size = 20),axis.title = element_text(size=15))+
      theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),panel.grid.major= element_blank(),panel.grid.minor = element_blank())
#"-Tissue.specific.Genes" for ordering from numbers of genes detected






#Heatmap
#Identification of "Tissue specific genes" amoung input genes

library(tidyr)
seExp<-output[[2]][["Lymph Node"]]
exp<-setNames(data.frame(assay(seExp), row.names = rowData(seExp)[,1]), colData(seExp)[,1])
exp$Gene<-row.names(exp)
exp<-exp %>% gather(key = "Tissue", value = "expression",1:(ncol(exp)-1))

ggplot(exp, aes(Tissue, Gene)) + geom_tile(aes(fill = expression),
     colour = "white") + scale_fill_gradient(low = "white",
     high = "steelblue")+
     labs(x='', y = '')+
      theme_bw()+
      guides(fill = guide_legend(title = "Log2(TPM)"))+
      #theme(legend.position="none")+
      theme(plot.title = element_text(hjust = 0.5,size = 20),axis.title = element_text(size=15))+
      theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),panel.grid.major= element_blank(),panel.grid.minor = element_blank())




```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
